### PR TITLE
Fix CI Visibility Gradle plugin to be compatible with Gradle build scans

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
# What Does This Do
Fixes the Gradle plugin injected by CI Visibility so that it does not cause Gradle build scans to fail.

# Additional Notes
Attaching tracer with CI Visibility to a Gradle build that has scanning enabled causes the following error (the code is obfuscated since build scans are part of Gradle Enterprise):
```
Gradle version: 8.6
23562Plugin version: 3.15.1
23563
23564com.gradle.scan.plugin.internal.operations.a: Build operation dispatch of started notification failed.
23565Operation context:
23566 org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$1 (1): {}
23567 org.gradle.configuration.BuildOperationFiringProjectsPreparer$ConfigureBuild$1 (104): {getBuildPath=:}
23568 org.gradle.initialization.buildsrc.BuildSourceBuilder$2$1 (169): {getBuildPath=:}
23569 org.gradle.configuration.BuildOperationFiringProjectsPreparer$ConfigureBuild$1 (178): {getBuildPath=:buildSrc}
23570 org.gradle.initialization.ProjectsEvaluatedNotifier$NotifyProjectsEvaluatedListeners$1 (293): {getBuildPath=:buildSrc}
23571 org.gradle.api.internal.plugins.DefaultPluginManager$OperationDetails (294): {pluginId=null, targetPath=:, pluginClass=datadog.trace.instrumentation.gradle.CiVisibilityPlugin, targetType=project, applicationId=32, buildPath=:buildSrc}
23572Caused by: java.lang.NullPointerException: Cannot invoke "java.net.URL.getProtocol()" because "<local4>" is null
23573 at com.gradle.scan.plugin.internal.c.ae.a.a.b.a(SourceFile:62)
23574 at com.gradle.scan.plugin.internal.c.ae.a.a.e.a(SourceFile:47)
23575 at com.gradle.scan.plugin.internal.c.ae.a.b.a(SourceFile:59)
23576 at com.gradle.scan.plugin.internal.c.ae.a.b.a(SourceFile:47)
23577 at com.gradle.scan.plugin.internal.operations.b$c.started(SourceFile:140)
23578 at com.gradle.scan.plugin.internal.operations.b.a(SourceFile:55)
23579 at com.gradle.scan.plugin.internal.operations.n.a(SourceFile:33)
23580 at com.gradle.scan.plugin.internal.operations.d.a(SourceFile:60)
23581 at com.gradle.scan.plugin.internal.operations.h.a(SourceFile:39)
23582 at com.gradle.scan.plugin.internal.r.a$a.a(SourceFile:31)
23583 at com.gradle.scan.plugin.internal.r.a$a.a(SourceFile:20)
23584 at com.gradle.scan.plugin.internal.r.a.c(SourceFile:67)
23585----------
```

The reason the error occurs is that Gradle build scan tries to determine the location of all the plugins that the build uses, including the plugin injected by CI Visibility.
The code is roughly:
```
ProtectionDomain domain = pluginClass.getProtectionDomain();
CodeSource codeSource = var2.getCodeSource();
if (codeSource != null) {
    URL locationUrl = var3.getLocation();
    String protocol = locationUrl.getProtocol(); // this is where the NPE happens
    // ...
}
```

The CI Visibility Gradle plugin is injected by the tracer as a helper class.
Currently all the helper classes are injected with no protection domain specified.
As the result, they get assigned `java.lang.ClassLoader#defaultDomain` - `new ProtectionDomain(new CodeSource(null, (Certificate[]) null), null, this, null)`: the default domain has a code source that returns `null` as its location, which causes the NPE above.

The proposed fix is to allow instrumentation modules to define the protection domain that will be assigned to the injected classes.

The protection domain for the Gradle plugin module has code source that points to the tracer JAR and class loader that is the CL where the plugin is injected.

Jira ticket: [SCV-157]

[SCV-157]: https://datadoghq.atlassian.net/browse/SCV-157?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by romainkomorndatadog — APPROVED** [review]

**Comment by PerfectSlayer** [comment]
I would like a review from @mcculls for this PR.

**Review by mcculls — CHANGES_REQUESTED** [review]
Have we submitted a report upstream to fix build scan? It looks like a bug in build scan to assume `locationUrl` is non-null

In terms of this PR I think we can simplify it a lot, we don't need the feature to supply arbitrary code-sources - we just need a flag to tell it to use the agent's code-source.

So I'd suggest removing the factory and have `HelperInjector` pick the domain based on the instrumentation's value of that flag.

**Comment by nikita-tkachenko-datadog** [comment]
> Have we submitted a report upstream to fix build scan? It looks like a bug in build scan to assume `locationUrl` is non-null
> 
> In terms of this PR I think we can simplify it a lot, we don't need the feature to supply arbitrary code-sources - we just need a flag to tell it to use the agent's code-source.
> 
> So I'd suggest removing the factory and have `HelperInjector` pick the domain based on the instrumentation's value of that flag.

[Issue](https://github.com/gradle/gradle/issues/29003) submitted.
I also reworked this PR to use the boolean flag instead of factory, as you suggested.

File: dd-java-agent/agent-builder/src/main/java/datadog/trace/agent/tooling/CombiningTransformerBuilder.java [hunk 1]
```diff
@@ -130,7 +130,10 @@ private void prepareInstrumentation(InstrumenterModule module, int instrumentati
     }
     helperTransformer =
         helperClassNames.length > 0
-            ? new HelperTransformer(module.getClass().getSimpleName(), helperClassNames)
+            ? new HelperTransformer(
+                module.injectHelperClassesWithAgentCodeSource(),
```
Comment (mcculls) [inline dd-java-agent/agent-builder/src/main/java/datadog/trace/agent/tooling/CombiningTransformerBuilder.java:134]: I think we can simplify this method name to just:
```suggestion
                module.useAgentCodeSource(),
```
and use javadoc to explain how/where it is used.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java [hunk 2]
```diff
@@ -135,12 +147,23 @@ private Map<String, Class<?>> injectClassLoader(
...
       INJECTING_HELPERS.end();
     }
   }
 
+  private ProtectionDomain createProtectionDomain(final ClassLoader classLoader) {
+    if (useAgentCodeSource) {
+      CodeSource codeSource = HelperInjector.class.getProtectionDomain().getCodeSource();
+      return new ProtectionDomain(codeSource, null, classLoader, null);
+    } else {
+      return null;
```
Comment (mcculls) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java:163]:
```suggestion
      return ClassLoadingStrategy.NO_PROTECTION_DOMAIN;
```
This constant boils down to the same literal, i.e. `null` but is more descriptive

**Review by mcculls — APPROVED** [review]
+1, thanks

just some minor suggestions

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java (hunk 2 above)
Comment (nikita-tkachenko-datadog) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/HelperInjector.java:163]: Good point, replaced literal with the constant

File: dd-java-agent/agent-builder/src/main/java/datadog/trace/agent/tooling/CombiningTransformerBuilder.java (hunk 1 above)
Comment (nikita-tkachenko-datadog) [reply dd-java-agent/agent-builder/src/main/java/datadog/trace/agent/tooling/CombiningTransformerBuilder.java:134]: Renamed the method

**Pull request merged**